### PR TITLE
[SPARK-47488][core][k8s] Driver stuck when thread pool is not shut down

### PR DESCRIPTION
**Pull request by littlelittlewhite09**
### What changes were proposed in this pull request?
The issue is open on JIRA. Link:[SPARK-47488](https://issues.apache.org/jira/browse/SPARK-47488)

### Why are the changes needed?
On k8s-client mode, if encounters exception and thread pool is not closed, driver pod may stuck. This pr is used to fix this issue.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
UT

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR, @littlelittlewhite09 .

Could you add a test case to Kubernetes Integration tests?

https://github.com/apache/spark/tree/master/resource-managers/kubernetes/integration-tests

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
...
           SparkContext.getActive.foreach(_.stop())
         } catch {
           case e: Throwable => logError(s"Failed to close SparkContext: $e")
-        }
+        } finally {
+          if (SparkContext.getActive.isEmpty) {
+            logWarning("Finished to close SparkContext inside driver pod successfully.")
+            if (!DriverPodIsNormal) {
+              logError(s"Driver Pod will exit because: $driverThrow")
+              System.exit(1)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1015]: Please use the official `SparkExitCode`.

https://github.com/apache/spark/blob/a3c04ec1145662e4227d57cd953bffce96b8aad7/core/src/main/scala/org/apache/spark/util/SparkExitCode.scala#L23

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
...
+          driverThrow = t
+        } else {
+          throw findCause(t)
+        }
     } finally {
       if (args.master.startsWith("k8s") && !isShell(args.primaryResource) &&
           !isSqlShell(args.mainClass) && !isThriftServer(args.mainClass) &&
           !isConnectServer(args.mainClass)) {
         try {
+          logWarning("Begin to close SparkContext inside driver pod......")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1006]: This extra warning message is a regression in a healthy Spark K8s jobs.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
         e
     }
 
+    //    this variable is used to judge whether driver pod is normal if spark is on k8s
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:987]: Extra spaces, `//      this`?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
...
           !isConnectServer(args.mainClass)) {
         try {
+          logWarning("Begin to close SparkContext inside driver pod......")
           SparkContext.getActive.foreach(_.stop())
         } catch {
           case e: Throwable => logError(s"Failed to close SparkContext: $e")
-        }
+        } finally {
+          if (SparkContext.getActive.isEmpty) {
+            logWarning("Finished to close SparkContext inside driver pod successfully.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1012]: This extra warning message is a regression in a healthy Spark K8s jobs.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
...
+            if (!DriverPodIsNormal) {
+              logError(s"Driver Pod will exit because: $driverThrow")
+              System.exit(1)
+            }
+          } else {
+            logWarning("Failed to close SparkContext.")
+          }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1019]: Indentation is correct? Line 1019 and 1020 seems to have some gap.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 6]
```diff
@@ -984,20 +984,39 @@ private[spark] class SparkSubmit extends Logging {
...
-        }
+        } finally {
+          if (SparkContext.getActive.isEmpty) {
+            logWarning("Finished to close SparkContext inside driver pod successfully.")
+            if (!DriverPodIsNormal) {
+              logError(s"Driver Pod will exit because: $driverThrow")
+              System.exit(1)
+            }
+          } else {
+            logWarning("Failed to close SparkContext.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1018]: May I ask when does this happen actually? And, what was the previous behavior?

**Comment by yaooqinn** [comment]
I checked your example code in the JIRA and it looks weird. Ideally, you should stop the non-daemon threads for a JVM to exit， and stop SparkContext to terminate the app

**Comment by littlelittlewhite09** [comment]
> I checked your example code in the JIRA and it looks weird. Ideally, you should stop the non-daemon threads for a JVM to exit， and stop SparkContext to terminate the app
If non-daemon threads is not stopped on yarn-cluster, driver can stop. But it does not work on k8s.

**Comment by littlelittlewhite09** [comment]
> I checked your example code in the JIRA and it looks weird. Ideally, you should stop the non-daemon threads for a JVM to exit， and stop SparkContext to terminate the app

When thread pool is not closed, driver can stop if app is on yarn-cluser. But it does not on k8s. I'm not sure if this situation is reasonable.

**Pull request closed without merging**